### PR TITLE
Cleanup static inline for fast paths and add configure option to control force inline

### PR DESCRIPTION
**Pull request by minsii**
## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by minsii** [comment]
test:mpich/ch4/most

**Comment by minsii** [comment]
test:mpich/ch4/most

File: src/mpl/include/mpl_base.h [hunk 1]
```diff
@@ -42,7 +42,11 @@
 #endif /* ATTRIBUTE */
 
 #define MPL_UNUSED ATTRIBUTE((unused))
+#ifdef MPL_ENABLE_FORCE_INLINE
 #define MPL_STATIC_INLINE_PREFIX ATTRIBUTE((always_inline)) static inline
+#else
+#define MPL_STATIC_INLINE_PREFIX static inline
+#endif
```
Comment (hzhou) [inline src/mpl/include/mpl_base.h:49]: Do we have a use case that we *don't want* always_inline?

Comment (shintaro-iwasaki) [reply src/mpl/include/mpl_base.h:49]: I have an opposite opinion: if the compiler doesn't want to inline, usually we should not inline.

Comment (hzhou) [reply src/mpl/include/mpl_base.h:49]: > I have an opposite opinion: if the compiler doesn't want to inline, usually we should not inline.

I think I can agree to that.

File: configure.ac [hunk 2]
```diff
@@ -415,9 +415,10 @@ AC_ARG_ENABLE(fast,
                    MPICHLIB_FFLAGS, and MPICHLIB_FCFLAGS. (default is -O2)
         ndebug   - Appends -DNDEBUG to MPICHLIB_CFLAGS.
         no-strict-alignment - relax strict alignment requirement
-        all|yes  - "O2" and "ndebug" are enabled
+        inline   - Force inline performance critical routines
+        all|yes  - "O2", "ndebug", and "force-inline" are enabled
         none     - None of above options, i.e. --disable-fast
-],,enable_fast=O2)
+],,[enable_fast=O2,inline])
```
Comment (hzhou) [inline configure.ac:421]: `--enable-fast=inline` is misleading. What we mean is `--enable-fast=always_inline`, and I think we should name it as what we mean.

I think I agree with @shintaro-iwasaki that it is questionable that we want to turn on "always inline" by default. 

Also, it is not obvious to me how do we *turn off* always inline? That is my reason that I prefer always inline turned off by default and `--enable-fast=all` or `--enable-fast=alwaysinline` turns it on.

File: src/mpl/include/mpl_base.h (hunk 1 above)
Comment (minsii) [reply src/mpl/include/mpl_base.h:49]: There are two issues in current main branch regarding static inline:
- Some functions use `static inline` directly, but some others use the wrapper macro `MPL_STATIC_INLINE_PREFIX `. They should be consistent as we don't distinguish functions that must be always_inline v.s. inline based on compiler's decision. 
- `MPL_STATIC_INLINE_PREFIX` always set `always_inline`. Thus for functions that use it, symbols cannot be generated for debugging even with `-O0 -g` (e.g., no debugging symbol for `MPIDI_put_unsafe`)

To fix these issues, I did two things in this PR:
- Replace direct `static inline` code with `MPL_STATIC_INLINE_PREFIX` in MPICH/ch4
- Add an option "inline" in --enable-fast to control `always_inline` in `MPL_STATIC_INLINE_PREFIX`. If "inline" is not given, the wrapper will be equal to `static inline`. By default, --enable-fast=O2,inline.

Do you think it makes sense?

Comment (minsii) [reply src/mpl/include/mpl_base.h:49]: To address @shintaro-iwasaki 's opinion, I agree with it too. So maybe we should keep the default value of --enable-fast as "O2" (i.e., "inline" is disabled by default)

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

test:mpich/custom
netmod: ch4:ofi
config: fast

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

test:mpich/custom
netmod: ch4:ofi
config: fast

**Review by hzhou — APPROVED** [review]
Looks good. Approved, pending on tests.

**Comment by minsii** [comment]
Failure in mpich-warnings/ucx seems irrelevant to this PR:
https://jenkins-pmrs.cels.anl.gov/job/mpich-warnings/2012/compiler=gcc-4,config=ch4-ucx,label=centos64_review/

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by minsii** [comment]
Locally verified the results with different --enable-fast configure options.

**Comment by hzhou** [comment]
@minsii Feel free to merge it whenever you are ready

**Comment by yfguo** [comment]
@minsii The STUBNM seems unaffected and still using `static inline` in this PR. Is this planned?

**Comment by minsii** [comment]
@yfguo Good catch! Will add a PR to replace `static inline` in stubnm.

**Pull request merged**